docs/examples.rst

```diff
@@ -0,0 +1,310 @@
+    stream = dataref.stream(start_offset=5, shuffle=True, shuffle_seed=777)
+
+    # Interpret the stream as a tensorflow dataset
+    records = yogadl.tensorflow.make_tf_dataset(stream)
+
+    # It's a real tf.data.Dataset; you can use normal tf.data operations on it.
+    batches = records.repeat(3).batch(5)
+
+    # (this part requires TensorFlow >= 2.0)
+    for batch in batches:
+        print(batch)
+
+This should print:
+
+.. code::
+
+    tf.Tensor([5 1 9 6 7], shape=(5,), dtype=int64)
+    tf.Tensor([1 7 3 9 8], shape=(5,), dtype=int64)
+    tf.Tensor([2 6 0 4 5], shape=(5,), dtype=int64)
+    tf.Tensor([9 5 3 0 8], shape=(5,), dtype=int64)
+    tf.Tensor([6 7 4 1 2], shape=(5,), dtype=int64)
+
+Notice that:
+
+ - The start_offset is only applied to the first epoch, so in this example
+   .repeat(3) gave us 2.5 epochs of data since we skipped the first epoch.
+ - The shuffle is a true shuffle. The shuffled stream samples from the whole
+   dataset without any concept of a "buffer".
+ - The shuffle is reproducible because we chose a shuffle seed.
+ - Each epoch is reshuffled.  Also the reshuffle is reproducible; if you set a
+   ``start_offset`` equal to twice the epoch length, your first epoch would be
+   the same as the third epoch of data without any ``start_offset``.
+
+
+Can I get the same features in fewer steps?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+As a matter of fact, you can! In order to support the common use-case of
+running the same dataset through many different models during model development
+or hyperparameter search, you can use the ``storage.cacheable()`` decorator to
+decorate a function that returns a datastet.
+
+When the decorated function is called the first time, it will run one time and
+save its output to ``storage``. On subsequent calls, the original function
+will not run, but its cached output will be returned instead.
+
+In this way, you can get the benefit of caching without a single script and
+only a single call against the ``storage`` object:
+
+.. code:: python
+
+    @storage.cacheable("range", "2.0")
+    def make_records():
+        print("Cache not found, making range v2 dataset...")
+        records = tf.data.Dataset.range(10).map(lambda x: 2*x)
+        return records
+
+    # Follow the same steps as before.
+    dataref = make_records()
+    stream = dataref.stream()
+    records = yogadl.tensorflow.make_tf_dataset(stream)
+    batches = records.repeat(3).batch(5)
+
+    for batch in batches:
+        print(batch)
+
+The ``storage.cacheble()`` decorator is multi-processing safe, so if two
+identical processes are configured to use the same storage, only one of them
+will create and save the dataset. The other one will wait for the dataset to
+be saved and will then read the dataset from the cache.
+
+
+End-to-end training example:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Here is an example of how you might use YogaDL to train on the second half of
```

Comment:
I think it might be worth simplifying this example a bit, and just start with an offset of 10k. I think L188-190 might distracting to users.

Reply:
How about:
```    # Stream the dataset starting halfway through it.
    stream = dataref.stream(
        start_offset=32000, shuffle=True, shuffle_seed=777
    )
```
?

Reply:
Perfect